src/N3Store.js

```diff
@@ -88,8 +88,7 @@ export default class N3Store {
   *_findInIndex(index0, key0, key1, key2, name0, name1, name2, graphId) {
     let tmp, index1, index2;
     // Depending on the number of variables, keys or reverse index are faster
-    const varCount = !key0 + !key1 + !key2,
-        entityKeys = varCount > 1 ? Object.keys(this._ids) : this._entities;
+    const entityKeys = !(key0 || key1 || key2) ? Object.keys(this._ids) : this._entities;
```

Comment:
I traced back the original reason for `varCount` to 287634e6755f51c9222070739e40ef448c042365 / https://github.com/rdfjs/N3.js/pull/61#issuecomment-220900112. However, this was 6 years ago. Have the circumstances changed (they very well might have)?

Reply:
Whilst I haven't had time to go through and understand the whole conversation, I would observe that `!graph` is always false (since `graph` is always true) and thus it would *always* use the `this._entities` equivalent in the original PR; so this is "going back" to that behaviour.

I'm actually not convinced that this ternary is necessary at all as I recall getting negligible performance changes if we *always* use `this._entities`

Reply:
> I'm actually not convinced that this ternary is necessary at all as I recall getting negligible performance changes if we always use this._entities.

Indeed it seems like this optimization makes no statistically significant change.

Current PR
```
jeswr@debian:~/jst/n/N3.js$ node perf/N3Store-perf.js 128
N3Store performance test
- Adding 2097152 triples to the default graph: 1.316s
* Memory usage for triples: 149MB
- Finding all 2097152 triples in the default graph 16384 times (0 variables): 5.690s
- Finding all 2097152 triples in the default graph 32768 times (1 variable): 1.871s
- Finding all 2097152 triples in the default graph 49152 times (2 variables): 1.919s

- Adding 1048576 quads: 870.469ms
* Memory usage for quads: 135MB
- Finding all 1048576 quads 131072 times: 1.405s

- Adding 1048576 with all different IRIs: 4.311s
* Retrieving all 1048576 quads: 1.350s
* Retrieving single by subject: 0.088ms
* Retrieving single by predicate: 0.071ms
* Retrieving single by object: 0.043ms
* Retrieving single by subject-predicate: 0.03ms
* Retrieving single by subject-object: 0.03ms
* Retrieving single by predicate-object: 0.053ms
* Retrieving single by subject-predicate-object: 0.097ms
```

Setting `const entityKeys = this._entities;` (which I suggest we do as this reduces memory requirements of matching against 3 variables.)
```
jeswr@debian:~/jst/n/N3.js$ node perf/N3Store-perf.js 128
N3Store performance test
- Adding 2097152 triples to the default graph: 1.233s
* Memory usage for triples: 149MB
- Finding all 2097152 triples in the default graph 16384 times (0 variables): 5.068s
- Finding all 2097152 triples in the default graph 32768 times (1 variable): 1.752s
- Finding all 2097152 triples in the default graph 49152 times (2 variables): 2.009s

- Adding 1048576 quads: 846.097ms
* Memory usage for quads: 124MB
- Finding all 1048576 quads 131072 times: 1.396s

- Adding 1048576 with all different IRIs: 5.237s
* Retrieving all 1048576 quads: 1.249s
* Retrieving single by subject: 0.056ms
* Retrieving single by predicate: 0.118ms
* Retrieving single by object: 0.068ms
* Retrieving single by subject-predicate: 0.058ms
* Retrieving single by subject-object: 0.072ms
* Retrieving single by predicate-object: 0.058ms
* Retrieving single by subject-predicate-object: 0.104ms
```

Reply:
The 0.xxxms tests should probably be increased to be at least a second each. Then we can see a little better. But the impact seems minimal indeed.

Reply:
The updated results are

`const entityKeys = !(key0 || key1 || key2) ? Object.keys(this._ids) : this._entities;`
```
N3Store performance test
- Adding 2097152 triples to the default graph: 1.182s
* Memory usage for triples: 149MB
- Finding all 2097152 triples in the default graph 16384 times (0 variables): 4.745s
- Finding all 2097152 triples in the default graph 32768 times (1 variable): 1.427s
- Finding all 2097152 triples in the default graph 49152 times (2 variables): 1.714s

- Adding 1048576 quads: 773.731ms
* Memory usage for quads: 131MB
- Finding all 1048576 quads 131072 times: 1.314s
N3 Store tests for sparsely connected entities
- Adding 1048576 with all different IRIs: 4.204s
* Retrieving all 1048576 quads: 1.946s
* Retrieving single by subject: 1.874s
* Retrieving single by predicate: 1.856s
* Retrieving single by object: 1.985s
* Retrieving single by subject-predicate: 2.176s
* Retrieving single by subject-object: 2.186s
* Retrieving single by predicate-object: 3.284s
* Retrieving single by subject-predicate-object: 2.449s
```

`const entityKeys = this._entities;`
```
N3Store performance test
- Adding 2097152 triples to the default graph: 1.220s
* Memory usage for triples: 147MB
- Finding all 2097152 triples in the default graph 16384 times (0 variables): 5.781s
- Finding all 2097152 triples in the default graph 32768 times (1 variable): 2.021s
- Finding all 2097152 triples in the default graph 49152 times (2 variables): 1.952s

- Adding 1048576 quads: 890.004ms
* Memory usage for quads: 123MB
- Finding all 1048576 quads 131072 times: 1.472s
N3 Store tests for sparsely connected entities
- Adding 1048576 with all different IRIs: 5.404s
* Retrieving all 1048576 quads: 1.214s
* Retrieving single by subject: 2.382s
* Retrieving single by predicate: 2.110s
* Retrieving single by object: 2.426s
* Retrieving single by subject-predicate: 2.598s
* Retrieving single by subject-object: 2.736s
* Retrieving single by predicate-object: 2.604s
* Retrieving single by subject-predicate-object: 2.662s
```

Reply:
So non-negligible, but not a unidirectional success.

Do you think we should change the detector to pick the optimal in all cases, or shall we just go with the most simple option?

Reply:
Although `Adding 1048576 with all different IRIs: 5.404s` bothers me; that should've been the same.

Reply:
> Adding 1048576 with all different IRIs: 5.404s bothers me

I re-ran this trial as a sanity check and got

```
N3Store performance test
- Adding 2097152 triples to the default graph: 1.187s
* Memory usage for triples: 146MB
- Finding all 2097152 triples in the default graph 16384 times (0 variables): 5.487s
- Finding all 2097152 triples in the default graph 32768 times (1 variable): 1.541s
- Finding all 2097152 triples in the default graph 49152 times (2 variables): 1.760s

- Adding 1048576 quads: 824.131ms
* Memory usage for quads: 131MB
- Finding all 1048576 quads 131072 times: 1.316s
N3 Store tests for sparsely connected entities
- Adding 1048576 with all different IRIs: 4.287s
* Retrieving all 1048576 quads: 1.092s
* Retrieving single by subject: 2.103s
* Retrieving single by predicate: 2.365s
* Retrieving single by object: 2.134s
* Retrieving single by subject-predicate: 2.450s
* Retrieving single by subject-object: 2.412s
* Retrieving single by predicate-object: 2.404s
* Retrieving single by subject-predicate-object: 2.371s
```

My machine is quite variable in the results (though I'm not sure why since I've only got firefox and vscode open). Even the changes that appear non-negligible appear to lie within the variance of my machine.

> Do you think we should change the detector to pick the optimal in all cases, or shall we just go with the most simple option?

I think we go with `const entityKeys = this._entities;` because it avoids introducing the overhead of creating an array when `Object.keys` is called so this means that the first results can still be retrieved faster *and* it takes up less memory

Reply:
That's great, let's do that.